src/java/org/apache/cassandra/index/sai/disk/v1/SSTableIndexWriter.java

```diff
@@ -213,18 +219,53 @@ else if (shouldFlush(sstableRowId))
             currentBuilder = newSegmentBuilder();
         }
 
-        if (term.remaining() == 0) return;
+        if (term.remaining() == 0)
+            return;
 
-        if (!TypeUtil.isLiteral(type))
+        updatesInFlight.incrementAndGet();
+        ForkJoinPool.commonPool().submit(() -> {
+            try
+            {
+                addTermInternal(term, key, sstableRowId, type);
```

Comment:
This method modifies mutable state within the `SegmentBuilder`. It doesn't look like the text or the numeric indexes support concurrent calls to add terms. Should we consider breaking this out as a vector only feature? It might only make sense to pay for the cost of context switches in the case of building the vector graph where it is likely more expensive to add an individual term.

Reply:
Generally SAI index builders are not thread safe, so even if `VectorSegmentBuilder` is thread safe, then this code violates the contract on `SegmentBuilder`. What about adding a new interface method `supportsConcurrentAdd` which would be true only for vector indexes and false by default? And then this code could decide how it wants to call add.

Reply:
good idea